<a href="https://colab.research.google.com/github/dakshini01/Statistical-Learning-e20181/blob/main/Assignment_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)

## 1. Bias of the Maximum Likelihood Estimator of Covariance

Consider the i.i.d. random variables

$$
X_1, X_2, \ldots, X_n \in \mathbb{R}^m
$$

with population mean

$$
\mu = E[X_i]
$$

and population covariance matrix

$$
\Sigma = Var(X_i).
$$

The empirical sample mean is

$$
\hat{\mu}_n = \frac{1}{n}\sum_{i=1}^{n}X_i.
$$

The raw Maximum Likelihood Estimator for the covariance matrix is

$$
\hat{\Sigma}_{MLE}
=
\frac{1}{n}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

However, this estimator is biased in finite samples. To show this, we use the identity

$$
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
=
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
-
n(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T.
$$

Taking expectation on both sides,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
E\left[
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
\right]
-
nE\left[
(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T
\right].
$$

Since

$$
E\left[
(X_i-\mu)(X_i-\mu)^T
\right]
=
\Sigma,
$$

we get

$$
E\left[
\sum_{i=1}^{n}
(X_i-\mu)(X_i-\mu)^T
\right]
=
n\Sigma.
$$

Also,

$$
Var(\hat{\mu}_n)
=
E\left[
(\hat{\mu}_n-\mu)(\hat{\mu}_n-\mu)^T
\right]
=
\frac{\Sigma}{n}.
$$

Therefore,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
n\Sigma
-
n\left(\frac{\Sigma}{n}\right).
$$

Hence,

$$
E\left[
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right]
=
(n-1)\Sigma.
$$

Now,

$$
E[\hat{\Sigma}_{MLE}]
=
E\left[
\frac{1}{n}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right].
$$

Therefore,

$$
E[\hat{\Sigma}_{MLE}]
=
\frac{1}{n}(n-1)\Sigma.
$$

Thus,

$$
E[\hat{\Sigma}_{MLE}]
=
\frac{n-1}{n}\Sigma.
$$

Since

$$
\frac{n-1}{n}\Sigma \neq \Sigma,
$$

the raw MLE covariance estimator is biased. It underestimates the true population covariance matrix.

---

## Bessel's Correction and the Unbiased Estimator

To remove this bias, we multiply the MLE covariance estimator by

$$
\frac{n}{n-1}.
$$

This gives the unbiased sample covariance matrix

$$
S
=
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

Taking expectation,

$$
E[S]
=
E\left[
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T
\right].
$$

Substituting the previous result,

$$
E[S]
=
\frac{1}{n-1}
(n-1)\Sigma.
$$

Therefore,

$$
E[S]=\Sigma.
$$

Hence, applying Bessel's correction gives an unbiased estimator of the population covariance matrix.

The reason for using $n-1$ instead of $n$ is that the sample mean $\hat{\mu}_n$ is estimated from the same dataset. Therefore, one degree of freedom is lost.

---

## 2. Type I Error and Type II Error in Structural Health Monitoring

In continuous structural health monitoring, the null hypothesis is

$$
H_0:
\text{The structural asset is healthy.}
$$

The alternative hypothesis is

$$
H_1:
\text{The structural asset is damaged or abnormal.}
$$

---

### Type I Error

A Type I Error occurs when the diagnostic system rejects $H_0$ even though $H_0$ is actually true.

In this context, it means that the system says the structure is damaged even though it is healthy.

Mathematically,

$$
\alpha
=
P(\text{Reject }H_0 \mid H_0 \text{ is true}).
$$

This is a false alarm.

In an industrial plant, a Type I Error may cause unnecessary shutdown, inspection cost, and production loss.

---

### Type II Error

A Type II Error occurs when the diagnostic system fails to reject $H_0$ even though $H_0$ is false.

In this context, it means that the system says the structure is healthy even though it is actually damaged.

Mathematically,

$$
\beta
=
P(\text{Fail to reject }H_0 \mid H_0 \text{ is false}).
$$

This is a missed detection.

A Type II Error is dangerous because real structural damage may continue unnoticed.

---

### Statistical Power

The statistical power of the diagnostic test is

$$
Power = 1-\beta.
$$

Power is the probability of correctly detecting damage when damage is truly present.

---

## Effect of an Ultra-Conservative Significance Threshold

Suppose the engineer chooses

$$
\alpha = 0.0001.
$$

This is a very small significance level.

The significance level $\alpha$ controls the probability of a false alarm. Therefore, by choosing a very small $\alpha$, the engineer greatly reduces the probability of incorrectly declaring a healthy structure as damaged.

However, this decision also makes the test more conservative.

The rejection region becomes very small:

$$
\alpha \downarrow
\Rightarrow
\text{Rejection region decreases}.
$$

As a result, it becomes harder for the test statistic to fall inside the rejection region.

Therefore,

$$
\beta \uparrow
$$

and

$$
Power = 1-\beta \downarrow.
$$

So, reducing $\alpha$ decreases false alarms, but it also reduces the ability of the diagnostic system to detect real damage.

---

## Geometric Interpretation Using the Healthy Confidence Ellipsoid

In a multivariate diagnostic system, healthy operation can be represented by a confidence ellipsoid around the baseline mean vector.

The ellipsoid represents the region where normal healthy data points are expected to lie.

When $\alpha$ is reduced from a common value such as

$$
\alpha = 0.05
$$

to an ultra-conservative value such as

$$
\alpha = 0.0001,
$$

the critical threshold increases.

Therefore, the healthy confidence ellipsoid becomes larger.

Geometrically,

$$
\alpha \downarrow
\Rightarrow
\text{critical value} \uparrow
\Rightarrow
\text{confidence ellipsoid volume} \uparrow.
$$

This means more observations are accepted as healthy.

Therefore:

- False alarms decrease.
- The healthy region becomes larger.
- The diagnostic system becomes less sensitive.
- Type II Error increases.
- Statistical power decreases.

Thus, an ultra-conservative threshold protects the plant from unnecessary shutdowns, but it increases the risk of missing subtle structural damage.

# Part B: Theoretical Extension (Asymptotic Distributions & Slutsky's Theorem)

The Lindeberg-Lévy Multivariate Central Limit Theorem states that, for i.i.d. random variables

$$
X_1, X_2, \ldots, X_n \in \mathbb{R}^m
$$

with population mean

$$
\mu = E[X_i]
$$

and population covariance matrix

$$
\Sigma = Var(X_i),
$$

the standardized sample mean satisfies

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,\Sigma)
$$

as

$$
n \rightarrow \infty.
$$

Here,

$$
\hat{\mu}_n = \frac{1}{n}\sum_{i=1}^{n}X_i.
$$

This means that when the sample size becomes very large, the distribution of the sample mean becomes approximately multivariate normal.

---

## Slutsky's Theorem

Let

$$
A_n \overset{d}{\longrightarrow} A
$$

and

$$
B_n \overset{p}{\longrightarrow} b,
$$

where $b$ is a constant.

Then Slutsky's Theorem states that

$$
A_n + B_n \overset{d}{\longrightarrow} A+b,
$$

$$
A_nB_n \overset{d}{\longrightarrow} Ab,
$$

and, if $b \neq 0$,

$$
\frac{A_n}{B_n}
\overset{d}{\longrightarrow}
\frac{A}{b}.
$$

In simple words, Slutsky's Theorem says that if one sequence converges in distribution and another sequence converges in probability to a constant, then we can combine them, and the final limiting distribution remains valid.

---

## Consistency of the Sample Covariance Matrix

The population covariance matrix $\Sigma$ is unknown in real applications.

Therefore, we estimate it using the unbiased sample covariance matrix

$$
S =
\frac{1}{n-1}
\sum_{i=1}^{n}
(X_i-\hat{\mu}_n)(X_i-\hat{\mu}_n)^T.
$$

By the Law of Large Numbers, the sample covariance matrix is a consistent estimator of the true covariance matrix. Therefore,

$$
S \overset{p}{\longrightarrow} \Sigma
$$

as

$$
n \rightarrow \infty.
$$

This means that when the sample size becomes very large, the sample covariance matrix $S$ becomes closer and closer to the true population covariance matrix $\Sigma$.

---

## Applying Slutsky's Theorem

From the Multivariate Central Limit Theorem,

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,\Sigma).
$$

But in practice, $\Sigma$ is unknown.

Since

$$
S \overset{p}{\longrightarrow} \Sigma,
$$

Slutsky's Theorem allows us to replace the unknown $\Sigma$ with the observable estimator $S$.

Therefore, the practical asymptotic distribution becomes

$$
\sqrt{n}(\hat{\mu}_n-\mu)
\overset{d}{\longrightarrow}
N(0,S).
$$

Equivalently, this can be written as

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right).
$$

---

## Interpretation

The expression

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right)
$$

means that the sample mean vector $\hat{\mu}_n$ is approximately multivariate normally distributed around the true population mean vector $\mu$.

The covariance of this distribution is

$$
\frac{S}{n}.
$$

As $n$ increases,

$$
\frac{S}{n}
\rightarrow 0.
$$

Therefore, the uncertainty around the sample mean becomes smaller.

This justifies using the sample mean and sample covariance matrix to build confidence regions, hypothesis tests, and diagnostic limits in practical structural health monitoring systems.

---

## Final Conclusion

Slutsky's Theorem justifies replacing the unknown population covariance matrix $\Sigma$ with the sample covariance matrix $S$ because

$$
S \overset{p}{\longrightarrow} \Sigma.
$$

Therefore, the theoretical model

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{\Sigma}{n}
\right)
$$

can be operationalized as

$$
\hat{\mu}_n
\sim
N\left(
\mu,
\frac{S}{n}
\right).
$$

This makes the asymptotic distribution practically usable because $S$ can be calculated directly from the observed dataset.

In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)

n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[
        [0.09, 0.02, 0.01],
        [0.02, 0.06, 0.03],
        [0.01, 0.03, 0.05]
    ],
    size=n_samples
)

# Inject subtle drift
base_data[4000:, 0] += 0.015
base_data[4000:, 2] -= 0.010

df_strain = pd.DataFrame(
    base_data,
    columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3']
)

# --------------------------------------------------------

def verify_first_moment_homogeneity(df, g_chunks=5):

    n = len(df)
    m = df.shape[1]

    chunk_size = n // g_chunks

    chunks = []

    for i in range(g_chunks):
        start = i * chunk_size
        end = (i + 1) * chunk_size
        chunks.append(df.iloc[start:end])

    # Global mean
    global_mean = df.mean().values.reshape(-1, 1)

    # Within-chunk matrix W
    W = np.zeros((m, m))

    # Between-chunk matrix B
    B = np.zeros((m, m))

    for chunk in chunks:

        chunk_values = chunk.values

        n_j = len(chunk)

        chunk_mean = chunk.mean().values.reshape(-1, 1)

        # Within matrix
        centered = chunk_values - chunk_mean.T

        W += centered.T @ centered

        # Between matrix
        mean_diff = chunk_mean - global_mean

        B += n_j * (mean_diff @ mean_diff.T)

    T = W + B

    wilks_lambda = np.linalg.det(W) / np.linalg.det(T)

    # Bartlett approximation
    chi_square = -(n - 1 - ((m + g_chunks) / 2)) * np.log(wilks_lambda)

    df_chi = m * (g_chunks - 1)

    p_value = 1 - stats.chi2.cdf(chi_square, df_chi)

    return {
        "Wilks Lambda": wilks_lambda,
        "Bartlett Chi Square": chi_square,
        "Degrees of Freedom": df_chi,
        "p-value": p_value
    }

results = verify_first_moment_homogeneity(df_strain, g_chunks=5)

for k, v in results.items():
    print(f"{k}: {v}")

Wilks Lambda: 0.9904519796929591
Bartlett Chi Square: 47.9215049961488
Degrees of Freedom: 12
p-value: 3.2255259508895406e-06


# Part C: Numerical Verification

The dataset was partitioned into

$$
g=5
$$

consecutive chunks, each containing

$$
n_j=1000
$$

observations.

The global mean vector and individual chunk mean vectors were calculated. Using these means, the Within-Chunk Sum of Squares and Cross Products Matrix

$$
W
$$

and the Between-Chunk Sum of Squares and Cross Products Matrix

$$
B
$$

were constructed.

The Wilks' Lambda statistic was computed as

$$
\Lambda
=
\frac{|W|}{|B+W|}.
$$

The calculated value was

$$
\Lambda = \text{0.9904519796929591}
$$

Using Bartlett's asymptotic transformation,

$$
\chi^2
=
-(n-1-\frac{m+g}{2})
\ln(\Lambda),
$$

the test statistic was obtained as

$$
\chi^2
=
\text{47.9215049961488}.
$$

The corresponding degrees of freedom were

$$
m(g-1)
=
3(5-1)
=
12.
$$

The resulting p-value was

$$
p
=
\text{3.2255259508895406e-06}.
$$

---

## Decision

The significance level was chosen as

$$
\alpha = 0.05.
$$

The decision rule is:

- If

$$
p \le 0.05
$$

reject

$$
H_0
$$

- If

$$
p > 0.05
$$

fail to reject

$$
H_0.
$$

Since

$$
p \text{ =3.2255259508895406e-06< } 0.05,
$$

we conclude that:


There is sufficient evidence to conclude that the chunk means differ significantly. Therefore, the baseline center has shifted over time and the first moment homogeneity assumption is violated.



# Geometric Subspace Optimization via Principal Component Analysis (PCA)

# Part A: Theoretical Fundamentals (Coordinate Projections & Orthogonality)

## 1. PCA Projection and Covariance Diagonalization

Let the centered baseline feature vector be

$$
\tilde{X}_i = X_i - \mu \in \mathbb{R}^m
$$

where $X_i$ is the original observation vector and $\mu$ is the population mean vector.

Since the data is centered,

$$
E[\tilde{X}_i] = 0.
$$

The population covariance matrix is

$$
\Sigma = E[\tilde{X}_i\tilde{X}_i^T].
$$

Because $\Sigma$ is symmetric and positive-semidefinite, it can be decomposed using spectral decomposition as

$$
\Sigma = P\Lambda P^T.
$$

Here,

$$
P = [p_1,p_2,\ldots,p_m]
$$

is the orthogonal eigenvector matrix, and

$$
\Lambda = diag(\lambda_1,\lambda_2,\ldots,\lambda_m)
$$

is the diagonal eigenvalue matrix.

The eigenvalues are ordered as

$$
\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_m \geq 0.
$$

Since $P$ is orthogonal,

$$
P^TP = PP^T = I.
$$

The principal component transformation is defined as

$$
Z_i = P^T\tilde{X}_i.
$$

Now we calculate the covariance matrix of the transformed vector $Z_i$:

$$
E[Z_iZ_i^T]
=
E[(P^T\tilde{X}_i)(P^T\tilde{X}_i)^T].
$$

Using the transpose rule,

$$
(AB)^T = B^TA^T,
$$

we get

$$
(P^T\tilde{X}_i)^T = \tilde{X}_i^TP.
$$

Therefore,

$$
E[Z_iZ_i^T]
=
E[P^T\tilde{X}_i\tilde{X}_i^TP].
$$

Since $P$ is constant, it can be moved outside the expectation:

$$
E[Z_iZ_i^T]
=
P^TE[\tilde{X}_i\tilde{X}_i^T]P.
$$

But

$$
E[\tilde{X}_i\tilde{X}_i^T] = \Sigma.
$$

So,

$$
E[Z_iZ_i^T]
=
P^T\Sigma P.
$$

Substituting the spectral decomposition,

$$
E[Z_iZ_i^T]
=
P^T(P\Lambda P^T)P.
$$

Using

$$
P^TP = I,
$$

we obtain

$$
E[Z_iZ_i^T]
=
I\Lambda I.
$$

Therefore,

$$
E[Z_iZ_i^T]
=
\Lambda.
$$

Hence, the covariance matrix of the transformed vector is diagonal.

---

## Interpretation of the Diagonal Result

Since

$$
E[Z_iZ_i^T] = \Lambda,
$$

we have

$$
\Lambda =
\begin{bmatrix}
\lambda_1 & 0 & \cdots & 0 \\
0 & \lambda_2 & \cdots & 0 \\
\vdots & \vdots & \ddots & \vdots \\
0 & 0 & \cdots & \lambda_m
\end{bmatrix}.
$$

This means all off-diagonal covariance terms are zero.

Therefore,

$$
E[Z_{i,j}Z_{i,k}] = 0
\quad
\text{for}
\quad
j \neq k.
$$

So, different principal component coordinates are uncorrelated.

Geometrically, PCA rotates the original coordinate system into a new coordinate system where the axes point in the directions of maximum variance.

Statistically, each principal component captures a separate uncorrelated mode of variation in the data.

---

## 2. Conservation of Total Variance

The total variance of the original system is measured by the trace of the covariance matrix:

$$
tr(\Sigma).
$$

Using spectral decomposition,

$$
\Sigma = P\Lambda P^T.
$$

Therefore,

$$
tr(\Sigma)
=
tr(P\Lambda P^T).
$$

Using the cyclic property of trace,

$$
tr(ABC)=tr(BCA)=tr(CAB),
$$

we get

$$
tr(P\Lambda P^T)
=
tr(\Lambda P^TP).
$$

Since

$$
P^TP = I,
$$

we get

$$
tr(\Sigma)
=
tr(\Lambda I).
$$

Therefore,

$$
tr(\Sigma)
=
tr(\Lambda).
$$

Since $\Lambda$ is diagonal,

$$
tr(\Lambda)
=
\sum_{j=1}^{m}\lambda_j.
$$

Thus,

$$
tr(\Sigma)
=
tr(\Lambda)
=
\sum_{j=1}^{m}\lambda_j.
$$

This proves that PCA preserves total variance.

PCA does not create or destroy variance. It only rotates the coordinate system.

---

## Cumulative Explained Variance Ratio

Suppose we retain only the first $k$ principal components, where

$$
k < m.
$$

The explained subspace is

$$
\hat{P}_k \in \mathbb{R}^{m \times k}.
$$

The residual unexplained subspace is

$$
\hat{P}_{m-k} \in \mathbb{R}^{m \times (m-k)}.
$$

The cumulative explained variance ratio is

$$
\Phi(k)
=
\frac{
\sum_{j=1}^{k}\hat{\lambda}_j
}{
\sum_{j=1}^{m}\hat{\lambda}_j
}.
$$

This tells us the percentage of total system variance captured by the first $k$ principal components.

---

## Residual Unexplained Variance Ratio

The residual unexplained variance ratio is

$$
\Psi(k)
=
1-\Phi(k).
$$

Equivalently,

$$
\Psi(k)
=
\frac{
\sum_{j=k+1}^{m}\hat{\lambda}_j
}{
\sum_{j=1}^{m}\hat{\lambda}_j
}.
$$

This tells us the percentage of system variance discarded by ignoring the remaining $m-k$ components.

---

## 3. Subspace Reconstruction and Residual Error

In structural health monitoring, we often keep only the first $k$ principal components.

The centered observation vector can be decomposed as

$$
x_i-\hat{\mu}_n
=
\hat{P}_kz_{i,k}
+
\hat{P}_{m-k}z_{i,m-k}.
$$

Here,

$$
\hat{P}_kz_{i,k}
$$

is the explained principal subspace component, and

$$
\hat{P}_{m-k}z_{i,m-k}
$$

is the residual subspace component.

The reconstructed vector is

$$
\hat{x}_i
=
\hat{\mu}_n
+
\hat{P}_kz_{i,k}.
$$

The residual error vector is

$$
e_i
=
x_i-\hat{x}_i.
$$

Substituting $\hat{x}_i$,

$$
e_i
=
x_i-
(\hat{\mu}_n+\hat{P}_kz_{i,k}).
$$

Since

$$
x_i-\hat{\mu}_n
=
\hat{P}_kz_{i,k}
+
\hat{P}_{m-k}z_{i,m-k},
$$

we get

$$
e_i
=
\hat{P}_{m-k}z_{i,m-k}.
$$

---

## Pythagorean Energy Decomposition

Since the principal subspace and residual subspace are orthogonal,

$$
\hat{P}_k^T\hat{P}_{m-k}=0.
$$

Therefore, the total squared Euclidean norm decomposes as

$$
\|x_i-\hat{\mu}_n\|^2
=
\|\hat{P}_kz_{i,k}\|^2
+
\|\hat{P}_{m-k}z_{i,m-k}\|^2.
$$

Because $\hat{P}_k$ and $\hat{P}_{m-k}$ are orthonormal bases,

$$
\|\hat{P}_kz_{i,k}\|^2
=
\|z_{i,k}\|^2
$$

and

$$
\|\hat{P}_{m-k}z_{i,m-k}\|^2
=
\|z_{i,m-k}\|^2.
$$

Thus,

$$
\|x_i-\hat{\mu}_n\|^2
=
\|z_{i,k}\|^2
+
\|z_{i,m-k}\|^2.
$$

This proves that the total energy of the centered observation is split into explained principal energy and unexplained residual energy.

---

## Hotelling's $T^2$ Statistic

Hotelling's $T^2$ statistic measures the distance of an observation inside the retained principal subspace.

It is defined as

$$
T_i^2
=
z_{i,k}^T
\hat{\Lambda}_k^{-1}
z_{i,k}.
$$

Equivalently,

$$
T_i^2
=
\sum_{j=1}^{k}
\frac{z_{i,j}^2}{\hat{\lambda}_j}.
$$

A large value of $T_i^2$ means that the observation is unusual within the normal major variation directions.

In structural health monitoring, $T^2$ is sensitive to abnormal changes that still occur inside the normal principal subspace.

For example, an environmental load anomaly such as extreme operational wind loads may increase the energy inside the main structural modes. Therefore, it is usually detected strongly by

$$
T^2.
$$

---

## Residual $Q$ Statistic / Squared Prediction Error

The residual $Q$ statistic, also called Squared Prediction Error, measures the energy left outside the retained principal subspace.

It is defined as

$$
Q_i
=
e_i^Te_i.
$$

Since

$$
e_i = \hat{P}_{m-k}z_{i,m-k},
$$

we get

$$
Q_i
=
z_{i,m-k}^Tz_{i,m-k}.
$$

Equivalently,

$$
Q_i
=
\sum_{j=k+1}^{m}
z_{i,j}^2.
$$

A large $Q_i$ means that the observation contains information that the retained PCA model cannot explain.

In structural health monitoring, $Q$ is sensitive to new fault patterns or structural changes outside the learned healthy subspace.

For example, an internal structural fracture such as a fatigue crack can change the structural load pathway and create new residual behavior. Therefore, it is usually detected strongly by

$$
Q.
$$

---

## Final Diagnostic Comparison

Hotelling's $T^2$ statistic monitors abnormal variation inside the retained principal subspace.

The residual $Q$ statistic monitors abnormal variation outside the retained principal subspace.

Therefore:

$$
T^2
\rightarrow
\text{anomaly inside the normal dominant operating modes}
$$

and

$$
Q
\rightarrow
\text{new unmodeled anomaly or structural fault outside the normal subspace}.
$$

For an environmental load anomaly such as extreme wind loading, the more sensitive statistic is usually

$$
T^2.
$$

For an internal structural fracture or fatigue crack, the more sensitive statistic is usually

$$
Q.
$$

In [2]:
import numpy as np
import pandas as pd

# ------------------------------------------------------------
# Synthetic baseline data generator
# ------------------------------------------------------------
np.random.seed(42)

n_samples = 3000
n_features = 4

mean = np.zeros(n_features)

cov = np.array([
    [1.00, 0.75, 0.45, 0.20],
    [0.75, 1.00, 0.50, 0.25],
    [0.45, 0.50, 1.00, 0.60],
    [0.20, 0.25, 0.60, 1.00]
])

X = np.random.multivariate_normal(mean, cov, size=n_samples)

df_baseline = pd.DataFrame(
    X,
    columns=["Sensor_1", "Sensor_2", "Sensor_3", "Sensor_4"]
)

# ------------------------------------------------------------
# PCA Optimization Pipeline
# ------------------------------------------------------------
class PCASubspaceOptimizer:

    def __init__(self, df):
        self.df = df.select_dtypes(include=[np.number])

    def analyze_subspaces(self):
        X = self.df.values

        n, m = X.shape

        # 1. Center and standardize data
        X_mean = np.mean(X, axis=0)
        X_std = np.std(X, axis=0, ddof=1)

        X_std[X_std == 0] = 1e-15

        Z = (X - X_mean) / X_std

        # 2. Unbiased sample covariance matrix
        S = np.cov(Z, rowvar=False, ddof=1)

        # 3. Eigen decomposition
        eigenvalues, eigenvectors = np.linalg.eigh(S)

        # Sort eigenvalues and eigenvectors in descending order
        idx = np.argsort(eigenvalues)[::-1]
        eigenvalues = eigenvalues[idx]
        eigenvectors = eigenvectors[:, idx]

        # 4. PCA scores
        scores = Z @ eigenvectors

        results = []

        # 5. Loop through k = 1, 2, ..., m-1
        for k in range(1, m):

            scores_k = scores[:, :k]
            lambda_k = eigenvalues[:k]

            residual_scores = scores[:, k:]

            # Hotelling's T²
            T2_values = np.sum((scores_k ** 2) / lambda_k, axis=1)

            # Q Statistic / SPE
            Q_values = np.sum(residual_scores ** 2, axis=1)

            mean_T2 = np.mean(T2_values)
            mean_Q = np.mean(Q_values)

            explained_variance_ratio = np.sum(eigenvalues[:k]) / np.sum(eigenvalues)
            residual_variance_ratio = 1 - explained_variance_ratio

            results.append({
                "k": k,
                "Mean T2": mean_T2,
                "Mean Q": mean_Q,
                "Explained Variance Ratio": explained_variance_ratio,
                "Residual Variance Ratio": residual_variance_ratio
            })

        return pd.DataFrame(results), eigenvalues, eigenvectors


# ------------------------------------------------------------
# Execute PCA optimizer
# ------------------------------------------------------------
pca_engine = PCASubspaceOptimizer(df_baseline)

pca_results, eigenvalues, eigenvectors = pca_engine.analyze_subspaces()

print("Eigenvalues:")
print(eigenvalues)

print("\nPCA Subspace Optimization Results:")
display(pca_results)

Eigenvalues:
[2.40261148 1.0067027  0.34290837 0.24777746]

PCA Subspace Optimization Results:


,k,Mean T2,Mean Q,Explained Variance Ratio,Residual Variance Ratio
0,1,0.999667,1.596856,0.600653,0.399347
1,2,1.999333,0.590489,0.852329,0.147671
2,3,2.999000,0.247695,0.938056,0.061944
